odo/backends/avro.py

```diff
@@ -0,0 +1,412 @@
+from __future__ import absolute_import, division, print_function
+
+import errno
+import os
+import uuid
+import json
+import fastavro
+import six
+from avro import schema, datafile, io
+from avro.schema import AvroException
+from multipledispatch import dispatch
+import pandas as pd
+from datashape import dshape, discover, var, Record, Map, Var, \
+    Option, null, string, int8, int32, int64, float64, float32, boolean, bytes_
+import datashape.coretypes as ct
+from collections import Iterator
+from ..append import append
+from ..convert import convert
+from ..resource import resource
+from ..temp import Temp
+
+try:
+    from avro.schema import make_avsc_object as schema_from_dict #Python 2.x
+except ImportError:
+    from avro.schema import SchemaFromJSONData as schema_from_dict  #Python 3.x
+
+PRIMITIVE_TYPES_MAP = {
+    'null': null,
+    'boolean': boolean,
+    'string': string,
+    'bytes': int8,
+    'int': int32,
+    'long': int64,
+    'double': float64,
+    'float': float32,
+}
+
+NAMED_TYPES_MAP = {
+    'fixed': bytes_, #TODO
+    'enum': int8, #TODO
+    'record': Record,
+    'error': Record, #TODO
+}
+
+COMPOUND_TYPES_MAP = {
+    'array': Var,
+    'map': Map,
+    'union': null, #TODO
+    'request': null, #TODO
+    'error_union': null, #TODO
+}
+
+
+AVRO_TYPE_MAP = {}
+AVRO_TYPE_MAP.update(PRIMITIVE_TYPES_MAP)
+AVRO_TYPE_MAP.update(NAMED_TYPES_MAP)
+AVRO_TYPE_MAP.update(COMPOUND_TYPES_MAP)
+
+dshape_to_avro_primitive_types = {
+    ct.int8: 'bytes',
+    ct.int16: 'int',
+    ct.int32: 'int',
+    ct.int64: 'long',
+    ct.float32: 'float',
+    ct.float64: 'double',
+    ct.date_: 'long',
+    ct.datetime_: 'long',
+    ct.string: 'string',
+    ct.bool_: 'boolean'
+}
+
+
+class AVRO(object):
+    """Wrapper object for reading and writing an Avro container file
+
+    Parameters
+    ----------
+
+    uri : str
+        uri of avro data
+
+    schema : avro.schema.Schema
+        User specified Avro schema object.  Used to decode file or serialize new records to file.
+         schema is required to create a new Avro file.
+         If reading or appending to an existing Avro file, the writers_schema embedded in that file
+         will be used.
+
+    codec : str
+        compression codec.  Valid values: 'null', 'deflate', 'snappy'
+
+    """
+    def __init__(self, path, schema=None, codec='null', **kwargs):
+        self._path = path
+        self._schema = schema
+        self._codec = codec
+
+    def __iter__(self):
+        return self.reader.__iter__()
+
+    def next(self):
+        return self.reader.next()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, value, traceback):
+        # Perform a close if there's no exception
```

Comment:
What about this case:

```python
with AVRO('path/to/file.avro', sch) as afh:
    for x in afh:
        print(x) 
    raise RuntimeError()
```

Won't the `if type is None` check leave our avro file unclosed, and wouldn't the user expect it to be closed even with an exception?